sdks/java/io/iceberg/src/main/java/org/apache/beam/io/iceberg/IcebergIOTranslation.java

```diff
@@ -0,0 +1,241 @@
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.util.construction.PTransformTranslation.TransformPayloadTranslator;
+import org.apache.beam.sdk.util.construction.SdkComponents;
+import org.apache.beam.sdk.util.construction.TransformPayloadTranslatorRegistrar;
+import org.apache.beam.sdk.values.Row;
+import org.apache.beam.vendor.grpc.v1p60p1.com.google.protobuf.ByteString;
+import org.apache.beam.vendor.guava.v32_1_2_jre.com.google.common.collect.ImmutableMap;
+import org.apache.iceberg.catalog.TableIdentifier;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+@SuppressWarnings({"rawtypes", "nullness"})
+public class IcebergIOTranslation {
+  static class IcebergIOReadTranslator implements TransformPayloadTranslator<ReadRows> {
+
+    static final Schema READ_SCHEMA =
+        Schema.builder()
+            .addByteArrayField("catalog_config")
+            .addNullableStringField("table_identifier")
+            .build();
+
+    public static final String ICEBERG_READ_TRANSFORM_URN =
+        "beam:transform:org.apache.beam:iceberg_read:v1";
+
+    @Override
+    public String getUrn() {
+      return ICEBERG_READ_TRANSFORM_URN;
+    }
+
+    @Override
+    public @Nullable FunctionSpec translate(
+        AppliedPTransform<?, ?, ReadRows> application, SdkComponents components)
+        throws IOException {
+      SchemaApi.Schema expansionSchema = SchemaTranslation.schemaToProto(READ_SCHEMA, true);
+      Row configRow = toConfigRow(application.getTransform());
+      ByteArrayOutputStream os = new ByteArrayOutputStream();
+      RowCoder.of(READ_SCHEMA).encode(configRow, os);
+
+      return FunctionSpec.newBuilder()
+          .setUrn(getUrn())
+          .setPayload(
+              SchemaAwareTransformPayload.newBuilder()
+                  .setExpansionSchema(expansionSchema)
+                  .setExpansionPayload(ByteString.copyFrom(os.toByteArray()))
+                  .build()
+                  .toByteString())
+          .build();
+    }
+
+    @Override
+    public Row toConfigRow(ReadRows transform) {
+
+      Map<String, Object> fieldValues = new HashMap<>();
+
+      if (transform.getCatalogConfig() != null) {
+        fieldValues.put("catalog_config", toByteArray(transform.getCatalogConfig()));
+      }
+      if (transform.getTableIdentifier() != null) {
+        TableIdentifier identifier = transform.getTableIdentifier();
+        List<String> identifierParts =
+            Arrays.stream(identifier.namespace().levels()).collect(Collectors.toList());
+        identifierParts.add(identifier.name());
+        fieldValues.put("table_identifier", String.join(".", identifierParts));
+      }
+
+      return Row.withSchema(READ_SCHEMA).withFieldValues(fieldValues).build();
+    }
+
+    @Override
+    public ReadRows fromConfigRow(Row configRow, PipelineOptions options) {
+      try {
+        ReadRows.Builder builder = new AutoValue_IcebergIO_ReadRows.Builder();
```

Comment:
I was expecting that we could build the transform using the schema-aware transform Row constructor instead of using a builder here.

Reply:
This translation is for the IO itself, not for its SchemaTransform implementation